homeassistant/components/stream/core.py

```diff
@@ -42,17 +64,69 @@ class Segment:
     duration: float = attr.ib(default=0)
     # For detecting discontinuities across stream restarts
     stream_id: int = attr.ib(default=0)
-    parts: list[Part] = attr.ib(factory=list)
+    # Parts are stored in a dict indexed by http range start for easy
+    # lookup when requested by byterange
+    # As of Python 3.7, insertion order is preserved, and we insert
+    # in sequential order, so the Parts are ordered
+    parts_by_http_range: dict[int, Part] = attr.ib(factory=dict)
     start_time: datetime.datetime = attr.ib(factory=datetime.datetime.utcnow)
+    # Store text of this segment's hls playlist for reuse
```

Comment:
Yes in general I think this partial generation/caching is necessary, because otherwise you have to iterate through every part of every segment on each playlist reload (potentially several times a second).
The way we are doing it here we only have to iterate over the parts that have not yet been rendered, and in general that is just 1 part.